help center: Update "Import from Mattermost".

drrosa opened the pull request:
Updates the help center article to match the style and formatting of [Import from Slack](https://zulip.com/help/import-from-slack) and replaces existing content with its corresponding Markdown macro.

**Remarks:**
Everything was kept consistent with the Slack article except that there's no "Clean up after the export" section as it doesn't seem necessary after Mattermost's export process.

The "Import details" section is missing a few bullet points compared to Slack's. Do we have information on the following?:
- Mattermost's user roles mapping to Zulip's user roles
- Messages in threads are imported, but they are not explicitly marked as being in a thread
- Message edit history and `@user joined #channel_name` messages are not imported

**Screenshots and screen captures:**
- https://zulip.com/help/import-from-mattermost
<img width="783" alt="image" src="https://user-images.githubusercontent.com/2343554/203471466-3cec365f-bf82-4071-bbc6-2c059565cd17.png">

<img width="645" alt="image" src="https://user-images.githubusercontent.com/2343554/203476915-6bd440d4-1bcf-4fd0-90ba-65bb16f39667.png">

<img width="458" alt="image" src="https://user-images.githubusercontent.com/2343554/203477062-d5741d07-770b-4342-8770-7aa1047b37ca.png">

<img width="675" alt="image" src="https://user-images.githubusercontent.com/2343554/203471641-663a5864-b82f-4e9e-b8d3-79617796dfd9.png">

<img width="414" alt="image" src="https://user-images.githubusercontent.com/2343554/203472083-9ec6c1de-a517-4c5c-b9db-b74157c412ce.png">

<img width="669" alt="image" src="https://user-images.githubusercontent.com/2343554/205465766-e65b0a19-b278-4880-a6bb-46572bdba699.png">

<img width="469" alt="image" src="https://user-images.githubusercontent.com/2343554/203471680-37604883-9ffb-4aef-9547-7bba7252cbaa.png">

<img width="627" alt="image" src="https://user-images.githubusercontent.com/2343554/203471782-d68b0dd4-d1c0-4fc6-bb3d-ea0e4bf4f5b3.png">

<img width="333" alt="image" src="https://user-images.githubusercontent.com/2343554/203471819-29959a1a-1364-4476-ae0a-17e58eff0fb6.png">

<details>
<summary>Self-review checklist</summary>

- [x] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability

Communicate decisions, questions, and potential concerns.

- [x] Explains differences from previous plans (e.g., issue description).
- [x] Calls out remaining decisions and concerns.
- [x] Automated tests

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/version-control.html)).

- [x] Each commit is a coherent idea.
- [x] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [x] Visual appearance of the changes.
- [x] Links.
</details>

drrosa commented:
@alya This PR is ready for your review.

mateuszmandera commented:
> * Mattermost's user roles mapping to Zulip's user roles


The logic seems to be just this in `process_user`:
```
    role = UserProfile.ROLE_MEMBER
    if is_team_admin(user_dict):
        role = UserProfile.ROLE_REALM_OWNER
```

so "Team administrator" (https://docs.mattermost.com/welcome/about-user-roles.html#team-admin) = Zulip org owner, and otherwise just a regular Zulip member?



> * Messages in threads are imported, but they are not explicitly marked as being in a thread

We set a general topic for everything `topic_name = "imported from mattermost"` - we certainly don't respect threads. Do we import messages that in threads? I suspect so - https://docs.mattermost.com/manage/bulk-export-tool.html doesn't particularly talk about threads or how to distinguish them in the exported data format, so I'd assume "thread" data is just additional data on top of the regular `Post` objects (which get converted to our Messages) that can be ignored? Not sure.


> * Message edit history and `@user joined #channel_name` messages are not imported

Message edit history doesn't appear to be imported indeed (see `process_raw_message_batch` - no such data gets passed to build_message). I don't know if there's an equivalent of those "user joined" messages in Mattermost, I suppose we can skip talking about this?

drrosa commented:
> so "Team administrator" (https://docs.mattermost.com/welcome/about-user-roles.html#team-admin) = Zulip org owner, and otherwise just a regular Zulip member?

Yeah, it appears so. But shouldn't it be something like: 

| Mattermost role              | Zulip role    |
|-------------------------|---------------|
| System Admin                  | Owner            |
| Team administrator         | Administrator |
| Member                            | Member          |
| Guest                                | Guest              |

> We set a general topic for everything `topic_name = "imported from mattermost"`

Okay so I think it would be accurate to say:
* Messages in threads are imported, but they are not explicitly marked as being in a thread

> Message edit history doesn't appear to be imported indeed (see `process_raw_message_batch` - no such data gets passed to build_message). I don't know if there's an equivalent of those "user joined" messages in Mattermost, I suppose we can skip talking about this

Yeah I think we should skip it and just say:
* Message edit history is not imported.

mateuszmandera commented:
> * Messages in threads are imported, but they are not explicitly marked as being in a thread
> Message edit history is not imported.

I think that would be right, yeah.

> Yeah, it appears so. But shouldn't it be something like:

Agreed that Guest status should be respected - perhaps we should open an issue to track implementing that? Not sure if your proposed admin mappings look right to me though - `System Admin` seems to be a special Mattermost role for the entire installation I think? And `Team Admin` is the "king" of their Team ("When a team is first created, the person who set it up is made a Team Admin") - and our equivalent of a Team is a Realm - and we convert each team into a separate realm?
```
def do_covert_data(....):
...
    for team in mattermost_data["team"]:
        realm_id = NEXT_ID("realm_id")
        team_name = team["name"]

        user_handler = UserHandler()
        subscriber_handler = SubscriberHandler()
        user_id_mapper = IdMapper()
        stream_id_mapper = IdMapper()
        huddle_id_mapper = IdMapper()

        print("Generating data for", team_name)
        realm = make_realm(realm_id, team)
```

The rest of the PR sounds good to me/

drrosa commented:
I updated the PR to address @mateuszmandera's feedback.

@alya ready for your review.

<img width="678" alt="image" src="https://user-images.githubusercontent.com/2343554/204944787-7a9541cb-38db-4f1e-8121-5fba1e11eab4.png">

alya commented:
Can we add info about how to do a Mattermost Cloud export?

From https://docs.mattermost.com/manage/cloud-data-export.html it sounds like you first have to set up a Mattermost server; I don't know if they have an easier way...

alya commented:
> We expect to add support for importing these data from Mattermost once Mattermost's export tool includes them.

Are we saying that none of the bullets are covered in Mattermost exports? Is that actually true?

We should probably just drop this sentence.

alya commented:
> Agreed that Guest status should be respected - perhaps we should open an issue to track implementing that?

Sounds good -- @mateuszmandera would you be up for filing this one?

mateuszmandera commented:
> > 
> 
> Sounds good -- @mateuszmandera would you be up for filing this one?

#23720

drrosa commented:
> Are we saying that none of the bullets are covered in Mattermost exports? Is that actually true?

No, that's a leftover sentence that I forgot about. Actually, I think it only applies to User avatars and Message edit history too  since Mattermost has not documented these data in their data export help articles. So maybe we should say something like:

> - Mattermost's export tool is incomplete and does not support exporting user
  avatars or message edit history. We expect to add support for importing these
  data from Mattermost once Mattermost's export tool includes them.
 
or should we keep only the first sentence and drop the second one?
@alya

drrosa commented:
> Can we add info about how to do a Mattermost Cloud export?
> From https://docs.mattermost.com/manage/cloud-data-export.html it sounds like you first have to set up a Mattermost server; I don't know if they have an easier way...

I think the server setup part is only relevant for the import process in the destination server. My understanding is that you only need their CLI tool installed locally to perform the export and to download the exported data.

drrosa commented:
> Can we add info about how to do a Mattermost Cloud export?

Would something like this actually work? 

<img width="474" alt="image" src="https://user-images.githubusercontent.com/2343554/205194905-49b5ca4c-ef12-4e2e-ab7c-91e8217f6716.png">

This should work assuming the `.zip` file contains a `.json` file equivalent to the output of the bulk export tool.
But maybe the command `mmctl extract run --json` from https://docs.mattermost.com/manage/mmctl-command-line-tool.html#mmctl-extract-run is also necessary?

@mateuszmandera do you know?

alya commented:
> or should we keep only the first sentence and drop the second one?

I would just go with that, yeah.

drrosa commented:
> We can shorten a bit to:
> 
> > Mattermost's export tool does not support exporting user
> > avatars or message edit history.
> 
> The "incomplete" is obvious from the rest of the sentence, and just sounds unnecessarily grumpy to me.

Yeah, I agree. We should scan for negative tone when updating help center docs, but this is probably the only instance I've seen so far. Hm, I will try to keep it in mind from now on; maybe I haven't been paying enough attention to this kind of thing.

I've updated the PR accordingly. Now we just need to figure out the cloud export stuff and #23720.

<img width="669" alt="image" src="https://user-images.githubusercontent.com/2343554/205465728-16f5cc1e-6c33-4281-b11b-02f789e16d85.png">

alya commented:
@mateuszmandera Would you be up for validating the cloud export instructions?

mateuszmandera commented:
Sure, will try testing these steps

mateuszmandera commented:
@drrosa @alya Okay, the instructions essentially work, though perhaps we should expand them - the `.zip` contains what we'd expect, but it's an `import.jsonl` filename. But the format seems to be the same as for our `export.json` files, so just renaming it works. So the steps I needed to do to fully test an import was:
1. Unzip the export.
2. Rename `import.jsonl` to `export.json`
3. Now, on this directory, you can successfully run the `manage.py convert_mattermost_data` command.

So I'm not sure, should the instructions be the way the currently are, or should we add the steps to unzip and rename the file?
We also have the whole `.tar.gz`-making step in other tabs `tar --transform 's|^|mattermost/|' -czf export.tar.gz data/ exported_emoji/ export.json`, but it feels perhaps redundant here?

alya commented:
Hmm, I see.

I'd like to keep things simple for users who are just going to email us the file anyway. Maybe we should provide alternative tabs for the self-hosted import for importing from a cloud instance vs. not? That way, we can just have the right set of commands to run for the two formats you might end up with.

drrosa commented:
I updated this PR in response to your feedback.

@mateuszmandera can you please have a look at the instructions in the screenshots below and let us know if they look good to you?

![image](https://user-images.githubusercontent.com/2343554/209737481-5e8e8b7f-e308-4eab-85e5-5d516ce14905.png)

![image](https://user-images.githubusercontent.com/2343554/209737540-012e77b0-1c79-45d0-a654-b40e591ac37c.png)

mateuszmandera commented:
@drrosa @alya What about something like this? (see my `[squash]` commit i pushed). Reasoning:
1. You can't run `convert_mattermost_data /tmp/export.json`, it needs a data directory, not just a .json file.
2. If we're making a new tab for `import from Cloud export`, then we can contain all the Cloud-export-specific instructions to that tab, so I deleted the conditional last point from the `export` part, and moved the fiddling with the `.jsonl` file to the `import` instructions.

drrosa commented:
> @drrosa @alya What about something like this? (see my `[squash]` commit i pushed). Reasoning:

Sounds good to me.

 I squashed your commit and also updated the command after "Alternatively, to import into a custom subdomain, run:".

**Updated screenshots:**
- Cloud instance export
<img width="465" alt="image" src="https://user-images.githubusercontent.com/2343554/210118765-72969059-f629-439e-9316-71cb7b0b158f.png">

- Self hosting (cloud export)
<img width="456" alt="image" src="https://user-images.githubusercontent.com/2343554/210118813-eacc7b63-6042-4cc4-a449-dcd6afc5bfe7.png">

alya commented:
Works for me!

@timabbott this PR has gone through a few rounds of review by @mateuszmandera and me -- please take a look!

timabbott commented:
So, this mostly works for me, but the line-wrapping of these code blocks is pretty bad/painful. One thing we should be doing is using extra leading whitespace when we're line-wrapping commands (and/or using shorter example path names to avoid excessive line-wrapping), but we should probably also have a conversation in #documentation about changing the default max width of code blocks in /help/.

timabbott commented:
Merged, thanks @drrosa!

drrosa commented:
> So, this mostly works for me, but the line-wrapping of these code blocks is pretty bad/painful. One thing we should be doing is using extra leading whitespace when we're line-wrapping commands (and/or using shorter example path names to avoid excessive line-wrapping), but we should probably also have a conversation in #documentation about changing the default max width of code blocks in /help/.

Yeah, good point. I created #23971 to track this issue.

alya commented:
I don't think an underspecified issue is going the best way to track this one... I started a discussion at https://chat.zulip.org/#narrow/stream/19-documentation/topic/line.20wrapping.20code.20blocks for how to handle line breaks, so that we can have a concrete solution for that part.

And any width adjustments I think should be part of a broader redesign pass, which we'll probably do in not too long.
